### ISSUE TITLE
How do I map stats back to participant?

### ISSUE THREAD

**Issue by ajorkowski** [twilio/twilio-video.js#817]
So when I do `room.getStats()` there seems to be an array of stats for each connected participant. How do I match up the stats to the participant? The only information we get is the `peerConnectionId` but I can't seem to find the internals that would give me the corresponding participant sid that corresponds to this connection. Is this even possible or am I not understanding the relationship from the stats to whatever? This is a peer-to-peer session.

 - [x] I have verified that the issue occurs with the latest twilio-video.js release and is not marked as a known issue in the [CHANGELOG.md](https://github.com/twilio/twilio-video.js/blob/master/CHANGELOG.md).
 - [x] I reviewed the [Common Issues](https://github.com/twilio/twilio-video.js/blob/master/COMMON_ISSUES.md) and open GitHub issues and verified that this report represents a potentially new issue.
 - [x] I verified that the [Quickstart](https://github.com/twilio/video-quickstart-js) application works in my environment.
 - [x] I am not sharing any [Personally Identifiable Information (PII)](https://www.twilio.com/docs/glossary/what-is-personally-identifiable-information-pii)
       or sensitive account information (API keys, credentials, etc.) when reporting this issue.

**Code to reproduce the issue:**

```js
room.getStats().then(stats => stats[0].peerConnectionId)
// How do I map a peerConnectionId back to a participant??
```

**Expected behavior:**

I would like to map the peerConnectionId back to a participant, so I can show the stats per-participant.

**Actual behavior:**

The only way I can map it back right now is use the `remoteVideoTrackStats` or `remoteAudioTrackStats` (if there are any) which has the track sid, and then map back from the room participants. If there are no remote tracks, and only local tracks, then I also cannot work out which participant this goes to.

**Software versions:**

 - [x] Browser(s): Chrome 78.0.3904.108
 - [x] Operating System: Windows 10
 - [x] twilio-video.js: 2.0.0-beta15
 - [x] Third-party libraries (e.g., Angular, React, etc.): React

**Comment by manjeshbhargav** [comment]
Hi @ajorkowski ,

Sorry for taking such a long time to respond. Each stats object has a Track SID associated with it. You can iterate the `tracks` collection for each Participant in `room.participants` and map the Track SID to the Participant.

Please let me know if this works for you.

Thanks,

Manjesh

**Comment by ajorkowski** [comment]
Hi @manjeshbhargav,

That does work (and this is my current work around), but this doesn't work if the remote participant(s) has not shared a track yet. Let's look at this example:

We have the (local) participant who is connected to 3 other (remote) participants, and is sharing a track. The remote participants have not shared any tracks yet. We will receive 3 stats reports from 'getstats()', each with a different `peerConnectionId`. Since only the local participant has shared a track, on the stats object only the `localAudioTrackStats`/`localVideoTrackStats` will have data, and `remoteAudioTrackStats`/`remoteVideoTrackStats` will be empty. This means using the Track SID I can only map back to the local participant (since I only have local tracks). How do I work out which stat object maps to which (remote) participant?

This is why I was hoping I could map the `peerConnectionId` back to a participant somehow, since this is the only unique id for each stats object in this situation. Ideally, I think it would be awesome if there was just a `participantId` on the stats object when I make this call.

Thanks so much for your time with this, its not a rush or anything since we have the workaround which most of the cases (considering our use case). Happy Holidays!

Cheers,
Felix

**Comment by manjeshbhargav** [comment]
Hi @ajorkowski ,

There is no one-to-one mapping between PeerConnections and Participants. For example, in a Group Room, there will be only one PeerConnection irrespective of how many Participants join the Room. But I still don't get why we would need to map Stats to RemoteParticipants when they haven't shared any Tracks yet (there won't be any Stats for that RemoteParticipant until they have shared at least one Track).

Thanks,

Manjesh

**Comment by ajorkowski** [comment]
Hi @manjeshbhargav,

True, this use case is only for peer-to-peer rooms... The general idea is that I can show the stats to/from each participant (ie the connection to remote participant 1 is good, but not great for participant 2). The moment the (local) participant starts sharing the track, `getStats()` starts returning values, and there are stats showing the `bytesSent`, `packetsSent` and `packetsLost` in the direction from the local participant to the remote participant (obviously no incoming data in this case, until the remote participant has added a track).

I think this can be really useful to map this to participants, because if the person hosting the session is sharing something, they can quickly see if any of the participants are having a bad connection (without the participants needing to share their own video/audio first etc).

Basically - if we are just looking at the remote stats, we are not seeing the full picture, because we need the stats of the outgoing video/audio too to form a full picture of the connection status. And these outgoing stats can happen before there are any incoming stats (but cannot be mapped unless the incoming stats are available).

Again - this is all probably an edge case because when we are using peer-to-peer we are looking at 1-1 or small groups, so most likely each participant will try to connect at least 1 track at some point, which means we can capture both the incoming/outgoing stats to/from each participant no problem.

Cheers,
Felix

**Comment by ajorkowski** [comment]
I'll just put this here for reference. This is what we get back from 'getStats()' (on Chrome) in a peer-to-peer room where the local participant has a track but the 2 connected remote participants do not. You can see the same local track is used but the stats are different:

```
[
  {
    "peerConnectionId": "a7810547-f401-4d2d-beb3-6210cb711ad5",
    "localAudioTrackStats": [],
    "localVideoTrackStats": [
      {
        "trackId": "f861a7ce-6688-49ac-bf5a-476758ed5d93",
        "trackSid": "MTc37c512f511364bebaea7f1d967a7465",
        "timestamp": 1577144324154,
        "ssrc": "2151848959",
        "packetsLost": null,
        "codec": "VP8",
        "bytesSent": 2671924,
        "packetsSent": 2515,
        "roundTripTime": null,
        "captureDimensions": null,
        "dimensions": {
          "width": 1280,
          "height": 720
        },
        "captureFrameRate": null,
        "frameRate": null
      }
    ],
    "remoteAudioTrackStats": [],
    "remoteVideoTrackStats": []
  },
  {
    "peerConnectionId": "e6d270fd-4f23-4dcd-8f81-317008ff171e",
    "localAudioTrackStats": [],
    "localVideoTrackStats": [
      {
        "trackId": "f861a7ce-6688-49ac-bf5a-476758ed5d93",
        "trackSid": "MTc37c512f511364bebaea7f1d967a7465",
        "timestamp": 1577144324155,
        "ssrc": "2809157213",
        "packetsLost": null,
        "codec": "VP8",
        "bytesSent": 321132,
        "packetsSent": 367,
        "roundTripTime": null,
        "captureDimensions": null,
        "dimensions": {
          "width": 640,
          "height": 360
        },
        "captureFrameRate": null,
        "frameRate": null
      }
    ],
    "remoteAudioTrackStats": [],
    "remoteVideoTrackStats": []
  }
]
```

Small tanget question: is there a reason the `framerate` stat is `null` on Chrome?

**Comment by manjeshbhargav** [comment]
@ajorkowski ,

If you are using Group Rooms, you can leverage the [Network Quality API](https://www.twilio.com/docs/video/using-network-quality-api) to estimate the network quality of all Participants.

In Peer-to-Peer Rooms, the stats related to remote media don't make sense if no Tracks are being shared by the RemoteParticipant. So, assuming at least one Track is being shared, then what you're already doing should cover most of your use cases.

Re: `framerate` being null on Chrome, I think that's because they have not updated their Stats implementation to match the latest W3C spec yet (for frame rate). Anyway, I'll take a look and see if we need to do something on our side.

I hope this answers your question. Please feel free to re-open the issue if you need more clarification.

Happy holidays!

Manjesh Malavalli
JSDK Team
